simphony_osp/session/session.py

```diff
@@ -126,6 +115,40 @@ def __bool__(self) -> bool:
     """See the docstring of `locked` for an explanation of what locking an
     environment means."""
 
+    _user_lock: bool = False
+    """See the docstring of `locked` for an explanation of what locking an
+    environment means.
+
+    This property manages locks performed by the user (by setting the
+    `locked` attribute).
+    """
+
+    def lock(self):
+        """Increase the lock count.
+
+        This way of locking is not meant to be used by end users, only
+        internally within SimPhoNy code, as it allows to lock the
+        environment several times, later requiring several unlocks, which is
+        unintuitive.
+
+        See the docstring of `locked` for an explanation of what locking an
+        environment means.
+        """
+        self._lock += 1
```

Comment:
This is indeed quite unintuitive. Why do you need to (un)lock multiple times?

Reply:
This is only meant to be used internally within SimPhoNy, where such a mechanism is needed. It cannot be defined as `_lock` because other classes have to call it. This is why you find some lines like this

```python
# ↓ --------------------- Public API --------------------- ↓ #
```

around the code. What is enclosed inside is what is meant to be called by an end-user. Only such functions will be documented in the [API Reference](https://simphony.readthedocs.io/en/latest/api_ref.html). I am actually frustrated with the fact that I have to include these functions not meant for the end user, because they will also show up if the user runs `help(Environment)`. If you have an idea of how to solve this (apart from constructing a whole new set of classes only for the user, which I have already thought of) I'd be very happy.

Reply:
The comment [here](https://github.com/simphony/osp-core/blob/1c7692944c085b4e5ae93d256a9755804d390b58/simphony_osp/interfaces/interface.py#L266) gives some insight.

> Session must not close when exiting its context manager (just in case the user enters the context manager again inside populate).

This means that we can get away with removing the internal lock if we assume the wrapper developer will be careful. I think in `populate` and `compute` there is not much value in using a context manager because the session for the base graph is already the default. Different is the case of `commit`, where `session_old`, `session_new` and `session_base` are available. 

Only if the user does stuff like this

```python
with self.session_old:
    old_attributes = {
        row[0]
        for row in sparql(f"""[...]""")
    }

with self.session_new:
    new_attributes = {
        row[0]
        for row in sparql(f"""[...]""")
    }

with self.session_base as base:  # redundant because it is already the default
    # do some logic, for example difference in amount of attributes
    difference = len(new_attributes) - len(old_attributes)

    self._engine.increase_attributes += difference
    base.get(oclass=ontology.Simulation).one()[ontology.num_attributes] = len(new_attributes)
```

It is a highly stupid example but hopefully shows the point: if we trust the developer not to use context managers in this way inside the wrapper methods then we can remove the internal lock. Otherwise we better leave it there.

Reply:
ok, let's leave it then